src/jit/gentree.h

```diff
@@ -4224,15 +4227,35 @@ struct GenTreeSIMD : public GenTreeJitIntrinsic
 struct GenTreeHWIntrinsic : public GenTreeJitIntrinsic
 {
     NamedIntrinsic gtHWIntrinsicId;
-
-    GenTreeHWIntrinsic(var_types type, GenTree* op1, NamedIntrinsic hwIntrinsicID, var_types baseType, unsigned size)
-        : GenTreeJitIntrinsic(GT_HWIntrinsic, type, op1, nullptr, baseType, size), gtHWIntrinsicId(hwIntrinsicID)
-    {
-    }
-
-    GenTreeHWIntrinsic(
-        var_types type, GenTree* op1, GenTree* op2, NamedIntrinsic hwIntrinsicID, var_types baseType, unsigned size)
-        : GenTreeJitIntrinsic(GT_HWIntrinsic, type, op1, op2, baseType, size), gtHWIntrinsicId(hwIntrinsicID)
+    int            gtNumArgs; // number of the non-imm arguments
+    int            gtIval;    // the imm8 argument
```

Comment:
Why do we need to carry these bits of info here, rather than in the lookup table?

Reply:
Originally, I planned to unify the code for fixed-imm and unfixed-imm intrinsics that carry the `ival` in this field and "op" pointers only point to reg/mem arguments (good to table-driven). However, now we seem to need to deal with "non-const" arguments by the "compiler fallback" that break this design. 

However, I still wonder that we may need to check constants and generate the exception/jump-table in the importer rather than hardcode in the codegen. If we decide to implement the checking in the importer, the codegen can be simplified.

Reply:
I guess I'm still not sure I understand.

We basically have the following cases:
* Instruction has a fixed, non-user provided, constant (such as `CompareEqual`)
  * We don't need to pass anything, it is looked up from the instruction table at emit (this is the same for direct and indirect calls). Nothing special is needed in the importer.
* Instruction has a user provided constant
  * In the importer, we check if its constant or in range and determine whether we can expand directly (is constant and is in range) or need to go through a GT_CALL (not constant)
  * In the codegen, if the value is constant, we expand directly; otherwise we emit a compiler fallback (failure when the value is not in range is done by the IL implementation currently and isn't something we need to do codegen for directly).

We should be able to unify both and distinguish them based on whether or not `ival` in the lookup table is `-1` (`(int)-1` indicates user-specified, otherwise `(byte)ival` is the fixed constant).

Reply:
Thanks for the good summary. 

> In the codegen, if the value is constant, we expand directly; otherwise we emit a compiler fallback.

Can we do this in the front-end (i.e., importer)?

> (failure when the value is not in range is done by the IL implementation currently and isn't something we need to do codegen for directly).

I do not think so. Generating the exception for out-of-range arguments in JIT is better because we may not want two range tables for each intrinsic (one in C#, one in JIT).

Reply:
> Can we do this in the front-end (i.e., importer)?

We can't emit the fallback in the importer, but we can detect if the value is constant and determine whether or not we should expand there. I also don't think we want to construct actual blocks for the fallback in the importer. This risks breaking based on future optimizations/etc.

See https://github.com/dotnet/coreclr/pull/15538/commits/a8d8aa3a4564628ff6c01b8bccbc6562e81bf7c3, for an example of this with `Sse.Shuffle`.

> I do not think so. Generating the exception for out-of-range arguments in JIT is better because we may not want two range tables for each intrinsic (one in C#, one in JIT).

My comment was mostly based on the current status quo. We've had lots of discussion about this (https://github.com/dotnet/corefx/issues/26194) but haven't landed concretely yet.

Reply:
> My comment was mostly based on the current status quo. We've had lots of discussion about this (dotnet/corefx#26194) but haven't landed concretely yet.

This discussion seems not related to the out-of-range exception. 
We have to check the valid argument's value in the JIT and generate the compiler fallback based-on the valid range of instructions. These two phases can share one "range-table", so I think JIT throwing exceptions would avoid much redundant work in C#.

Reply:
The other exceptions (those that aren't PNSE) were discussed starting [here](https://github.com/dotnet/corefx/issues/26194#issuecomment-356387606). It would require additional JIT helper functions to throw the correct exceptions with the correct messages.

There wasn't a clear consensus on whether or not we should add new JIT helpers (so we could throw these exceptions from the JIT) or if we should do the existing `ThrowHelper` pattern in IL (https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx.cs#L235).

Reply:
> There wasn't a clear consensus on whether or not we should add new JIT helpers (

We already have the JIT-helper for this exception that is called `CORINFO_HELP_THROW_ARGUMENTOUTOFRANGEEXCEPTION`.

Reply:
Ah, ok. Then I see no reason not to do that one from the JIT (and I think that means we only don't have one for the `this type is not supported` exception, which might push the discussion towards consensus)

Reply:
The only problem which I have encountered is that **JIT helpers do not provide support for custom messages** what would be very helpful for intrinsics. If we could refactor helpers to support it, providing it is possible, than some other problems with exception throwing would be solved as well.